crates/wasi-common/src/host.rs

```diff
@@ -39,10 +39,25 @@ impl FileType {
     pub(crate) fn to_wasi(&self) -> __wasi_filetype_t {
         *self as __wasi_filetype_t
     }
+
+    pub(crate) fn from_wasi(wasi_filetype: u8) -> Option<Self> {
+        use FileType::*;
+        match wasi_filetype {
+            __WASI_FILETYPE_UNKNOWN => Some(Unknown),
+            __WASI_FILETYPE_BLOCK_DEVICE => Some(BlockDevice),
+            __WASI_FILETYPE_CHARACTER_DEVICE => Some(CharacterDevice),
+            __WASI_FILETYPE_DIRECTORY => Some(Directory),
+            __WASI_FILETYPE_REGULAR_FILE => Some(RegularFile),
+            __WASI_FILETYPE_SOCKET_DGRAM => Some(SocketDgram),
+            __WASI_FILETYPE_SOCKET_STREAM => Some(SocketStream),
+            __WASI_FILETYPE_SYMBOLIC_LINK => Some(Symlink),
+            _ => None,
+        }
+    }
 }
 
 #[derive(Debug, Clone)]
-pub(crate) struct Dirent {
+pub struct Dirent {
```

Comment:
How come `Dirent` needs to be `pub` now? Is it because when using `VirtualFile`, `Dirent` is now client-facing?

Reply:
Indeed, `readdir` does us in:
```
    fn readdir(
        &self,
        _cookie: wasi::__wasi_dircookie_t,
    ) -> Result<Box<dyn Iterator<Item = Result<Dirent>>>> {
        Err(Error::EBADF)
    }
```

How much do we want to avoid a `pub Dirent`? Does it makes sense to have a different `virtfs`-specific struct with an `Into<Dirent>` impl instead?

Reply:
Hmm, I guess that's fine then. I'd prefer to have only `Dirent` struct shared between the two, but that also depends on the exact scope of virtual fs. I think leaving `pub struct Dirent` is far for now. We can always refactor this later!